src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Semantic/Services/DefaultRazorSemanticTokensInfoService.cs

```diff
@@ -401,17 +377,46 @@ internal class DefaultRazorSemanticTokensInfoService : RazorSemanticTokensInfoSe
             return new VersionedSemanticRange(result, csharpResponse.ResultId, csharpResponse.IsFinalizedCSharp);
         }
 
+        // Internal for testing only
+        internal static bool TryGetMinimalCSharpRange(RazorCodeDocument codeDocument, [NotNullWhen(true)] out Range? range)
+        {
+            var csharpDoc = codeDocument.GetCSharpDocument();
+
+            // If there aren't any source mappings, there's no C# code in the Razor doc.
+            if (csharpDoc.SourceMappings.Count == 0)
+            {
+                range = null;
+                return false;
+            }
+
+            // We only need to colorize the portions of the generated doc that correspond with a C# mapping.
+            // To accomplish this while minimizing the amount of work we need to do, we'll only colorize the
+            // range spanning from the first C# span to the last C# span.
+            var csharpSourceText = codeDocument.GetCSharpSourceText();
+            var start = csharpSourceText.Lines.GetLinePosition(csharpDoc.SourceMappings[0].GeneratedSpan.AbsoluteIndex);
+            var startPosition = new Position(start.Line, start.Character);
+
+            var lastCSharpSpan = csharpDoc.SourceMappings[csharpDoc.SourceMappings.Count - 1].GeneratedSpan;
```

Comment:
I assume that `SourceMappings` is known to be an ordered array? Makes sense that it would be but I've never been in a position to verify it.